### PR TITLE
metrics: Add canceled requests

### PR DESCRIPTION
**Pull request by klauspost**
## Description

Add metric for canceled requests

## Motivation and Context

Better visibility for client cancellation.

## How to test this PR?

Observe metrics, cancel requests.

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

### CODE REVIEW

**Review by aead — APPROVED** [review]

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#62654b90-8cdd-4f33-b5e6-7bbc9d312d8a) |
<a name="62654b90-8cdd-4f33-b5e6-7bbc9d312d8a"></a>
### 11881-b3fd05e/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:30442
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 051617da7a0a:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 44 seconds
(4/15) Running aws-sdk-ruby tests ... done in 5 seconds
(5/15) Running awscli tests ... FAILED in 32 seconds
{
  "name": "awscli",
  "duration": 2592,
  "function": "aws --endpoint-url http://minio-c3.minio.io:30442 s3api copy-object --bucket awscli-mint-test-bucket-24381 --key datafile-1-kB-copy --copy-source awscli-mint-test-bucket-24381/datafile-1-kB\n",
  "status": "FAIL",
  "error": "Hash mismatch expected 084e1383b70fb0c51acc680fef370023, got ac57de7156d7fc25ac1a65f81fa3989b"
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 47 seconds
(7/15) Running minio-dotnet tests ... done in 37 seconds
(8/15) Running minio-go tests ... FAILED in 2 minutes and 35 seconds
{
  "args": {},
  "duration": 288,
  "error": "At least one of the pre-conditions you specified did not hold",
  "function": "CopyObjectPart(destination, source)",
  "message": "CopyObjectPart call failed",
  "name": "minio-go: testUnencryptedToSSES3CopyObjectPart",
  "status": "FAIL"
}
(8/15) Running minio-java tests ... FAILED in 2 minutes and 3 seconds
{
  "name": "minio-java",
  "function": "copyObject()",
  "args": "[match etag]",
  "duration": 347,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = PreconditionFailed, message = At least one of the pre-conditions you specified did not hold, bucketName = minio-java-test-390laj0, objectName = minio-java-test-k924bb-copy, resource = /minio-java-test-390laj0/minio-java-test-k924bb-copy, requestId = 166F43FEC63B1C37, hostId = 53f90bab-9dfd-4a5f-a987-c8e4fd9f2e2d)\nrequest={method=PUT, url=http://minio-c3.minio.io:30442/minio-java-test-390laj0/minio-java-test-k924bb-copy, headers=x-amz-copy-source-if-match: 71cff0a060f852067e443ad1e24ae26c-1\nx-amz-copy-source: /minio-java-test-1eabtd1/minio-java-test-k924bb\nHost: minio-c3.minio.io:30442\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210324T113232Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210324/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-copy-source;x-amz-copy-source-if-match;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=412, headers=Accept-Ranges: bytes\nContent-Length: 416\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nETag: \"71cff0a060f852067e443ad1e24ae26c\"\nLast-Modified: Wed, 24 Mar 2021 11:32:32 GMT\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 166F43FEC63B1C37\nX-Xss-Protection: 1; mode=block\nDate: Wed, 24 Mar 2021 11:32:32 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.execute(MinioClient.java:563), io.minio.MinioClient.executePut(MinioClient.java:904), io.minio.MinioClient.copyObject(MinioClient.java:1232), FunctionalTest.testCopyObjectMatchETag(FunctionalTest.java:1850), FunctionalTest.copyObject(FunctionalTest.java:2016), FunctionalTest.runObjectTests(FunctionalTest.java:3757), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(8/15) Running minio-js tests ... done in 56 seconds
(9/15) Running minio-py tests ... done in 3 minutes and 42 seconds
(10/15) Running s3cmd tests ... FAILED in 5 seconds
{
  "name": "s3cmd",
  "duration": "2899",
  "function": "test_put_object_multipart",
  "status": "FAIL",
  "error": "WARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Too many failures. Giving up on '/mint/data/datafile-65-MB'\nERROR: \nUpload of '/mint/data/datafile-65-MB' part 1 failed. Use\n  /usr/local/bin/s3cmd abortmp s3://s3cmd-test-bucket-5239/s3cmd-test-object-28951 378d6c07-4126-44d2-a97f-8efc0e3f1bee\nto abort the upload, or\n  /usr/local/bin/s3cmd --upload-id 378d6c07-4126-44d2-a97f-8efc0e3f1bee put ...\nto continue the upload.\nERROR: Upload of '/mint/data/datafile-65-MB' failed too many times (Last reason: )"
}
(10/15) Running s3select tests ... done in 10 seconds
(11/15) Running security tests ... done in 0 seconds

Executed 11 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Review by vadmeste — CHANGES_REQUESTED** [review]
v2 metrics doesn't seem to work for me

File: cmd/http-stats.go [hunk 1]
```diff
@@ -175,6 +179,11 @@ func (st *HTTPStats) updateStats(api string, r *http.Request, w *logger.Response
 		if !successReq && w.StatusCode != 0 {
 			st.totalS3Errors.Inc(api)
 		}
+		select {
+		case <-r.Context().Done():
```
Comment (aead) [inline cmd/http-stats.go:184]: Maybe check if the request actually got cancelled:
```go
if err := r.Context().Err(); err == context.Canceled {

}
```

**Comment by klauspost** [comment]
@vadmeste 
```
$ curl 127.0.0.1:9001/minio/v2/metrics/node
...
# HELP minio_s3_requests_canceled_total Total number S3 requests that were canceled from the client while processing
# TYPE minio_s3_requests_canceled_total counter
minio_s3_requests_canceled_total{api="putobject",server="127.0.0.1:9001"} 10
# HELP minio_s3_requests_error_total Total number S3 requests with errors
# TYPE minio_s3_requests_error_total counter
minio_s3_requests_error_total{api="putobject",server="127.0.0.1:9001"} 6
# HELP minio_s3_requests_inflight_total Total number of S3 requests currently in flight
# TYPE minio_s3_requests_inflight_total gauge
minio_s3_requests_inflight_total{api="getbucketlocation",server="127.0.0.1:9001"} 0
minio_s3_requests_inflight_total{api="getbucketversioning",server="127.0.0.1:9001"} 0
minio_s3_requests_inflight_total{api="headbucket",server="127.0.0.1:9001"} 0
minio_s3_requests_inflight_total{api="putobject",server="127.0.0.1:9001"} 0
# HELP minio_s3_requests_total Total number S3 requests
# TYPE minio_s3_requests_total counter
minio_s3_requests_total{api="getbucketlocation",server="127.0.0.1:9001"} 1
minio_s3_requests_total{api="getbucketversioning",server="127.0.0.1:9001"} 1
minio_s3_requests_total{api="headbucket",server="127.0.0.1:9001"} 1
minio_s3_requests_total{api="putobject",server="127.0.0.1:9001"} 45
...
```
Seemed to appear after a while??

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#11900897-3a76-4aba-9db7-f58727918beb) |
<a name="11900897-3a76-4aba-9db7-f58727918beb"></a>
### 11881-a3d6bb2/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c2.minio.io:30853
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp de57801813e7:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 2 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 44 seconds
(4/15) Running aws-sdk-ruby tests ... done in 5 seconds
(5/15) Running awscli tests ... FAILED in 33 seconds
{
  "name": "awscli",
  "duration": 3314,
  "function": "aws --endpoint-url http://minio-c2.minio.io:30853 s3api copy-object --bucket awscli-mint-test-bucket-29550 --key datafile-1-kB-copy --copy-source awscli-mint-test-bucket-29550/datafile-1-kB\n",
  "status": "FAIL",
  "error": "Hash mismatch expected 084e1383b70fb0c51acc680fef370023, got ac57de7156d7fc25ac1a65f81fa3989b"
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 47 seconds
(7/15) Running minio-dotnet tests ... done in 56 seconds
(8/15) Running minio-go tests ... FAILED in 2 minutes and 4 seconds
{
  "args": {},
  "duration": 742,
  "error": "At least one of the pre-conditions you specified did not hold",
  "function": "CopyObjectPart(destination, source)",
  "message": "CopyObjectPart call failed",
  "name": "minio-go: testUnencryptedToSSES3CopyObjectPart",
  "status": "FAIL"
}
(8/15) Running minio-java tests ... FAILED in 1 minutes and 49 seconds
{
  "name": "minio-java",
  "function": "copyObject()",
  "args": "[match etag]",
  "duration": 297,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = PreconditionFailed, message = At least one of the pre-conditions you specified did not hold, bucketName = minio-java-test-1rlh7c4, objectName = minio-java-test-232o0o3-copy, resource = /minio-java-test-1rlh7c4/minio-java-test-232o0o3-copy, requestId = 166F52F8C60EF472, hostId = 6031dc53-cf71-4168-a168-76cd9ed340df)\nrequest={method=PUT, url=http://minio-c2.minio.io:30853/minio-java-test-1rlh7c4/minio-java-test-232o0o3-copy, headers=x-amz-copy-source-if-match: 71cff0a060f852067e443ad1e24ae26c-1\nx-amz-copy-source: /minio-java-test-n59lnk/minio-java-test-232o0o3\nHost: minio-c2.minio.io:30853\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210324T160659Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210324/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-copy-source;x-amz-copy-source-if-match;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=412, headers=Accept-Ranges: bytes\nContent-Length: 418\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nETag: \"71cff0a060f852067e443ad1e24ae26c\"\nLast-Modified: Wed, 24 Mar 2021 16:06:59 GMT\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 166F52F8C60EF472\nX-Xss-Protection: 1; mode=block\nDate: Wed, 24 Mar 2021 16:06:59 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.execute(MinioClient.java:563), io.minio.MinioClient.executePut(MinioClient.java:904), io.minio.MinioClient.copyObject(MinioClient.java:1232), FunctionalTest.testCopyObjectMatchETag(FunctionalTest.java:1850), FunctionalTest.copyObject(FunctionalTest.java:2016), FunctionalTest.runObjectTests(FunctionalTest.java:3757), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(8/15) Running minio-js tests ... done in 49 seconds
(9/15) Running minio-py tests ... done in 3 minutes and 30 seconds
(10/15) Running s3cmd tests ... FAILED in 6 seconds
{
  "name": "s3cmd",
  "duration": "3017",
  "function": "test_put_object_multipart",
  "status": "FAIL",
  "error": "WARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Too many failures. Giving up on '/mint/data/datafile-65-MB'\nERROR: \nUpload of '/mint/data/datafile-65-MB' part 1 failed. Use\n  /usr/local/bin/s3cmd abortmp s3://s3cmd-test-bucket-28437/s3cmd-test-object-28865 64791184-ed7d-4271-bc71-738abf2909e3\nto abort the upload, or\n  /usr/local/bin/s3cmd --upload-id 64791184-ed7d-4271-bc71-738abf2909e3 put ...\nto continue the upload.\nERROR: Upload of '/mint/data/datafile-65-MB' failed too many times (Last reason: )"
}
(10/15) Running s3select tests ... done in 11 seconds
(11/15) Running security tests ... done in 0 seconds

Executed 11 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Review by vadmeste — APPROVED** [review]
LGTM

**Pull request merged**